src/platform_impl/linux/wayland/window/state.rs

```diff
@@ -155,6 +152,12 @@ pub struct WindowState {
 
     /// The underlying SCTK window.
     pub window: Window,
+
+    // NOTE: The spec says that destroying parent(`window` in our case), will unmap the
+    // subsurfaces. Thus to achieve atomic unmap of the client, drop the decorations
+    // frame after the `window` is dropped.
+    /// The window frame, which is created from the configure request.
+    frame: Option<WinitFrame>,
```

Comment:
Instead of moving it around, why not just add this to the `drop` method for this structure?

```rust
fn drop(&mut self) {
    self.frame = None;
}
```

Reply:
You need a manual drop, to drop the `window` first.

Reply:
Then we should wrap the `window` in an `Option` as well. I think it is fragile to rely on the field ordering here, as I believe the Rust team has said that it is not stable

Reply:
Then you have to wrap all the objects like that.

Reply:
Drop order is explicitly defined here, so I don't see a reason to use it. https://doc.rust-lang.org/reference/destructors.html

Reply:
Zero-cost option would be to use `ManuallyDrop`, but the standard library [explicitly recommends against that](https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html#manuallydrop-and-drop-order), so I think that relying on drop order is the cleaner option here; it's subtle, but it's documented, safe and stable.

Reply:
I think drop order is something that can easily be missed, so at an absolute minimum "drop order" should specifically be mentioned in the comment.

Reply:
Should I really document how every bit in rust works, though? If it has a `NOTE:` comment it's already raising attention to at least look into `blame` on why it's like that.

Reply:
I think a few words for a pretty unknown and non-obvious part of the specification is appropriate here.

Reply:
Yeah drop order is probably the issue